Refactored out into the Shared/Tests solution directory

MarkMichaelis opened the pull request:
Refactored out a file into the Shared/Tests solution directory that may be used to execute PowerShell scripts.  This is used to execute the HelloWorldInC#.ps1 file.

Fixes #209

MarkMichaelis committed:
Added some test generated directories that should be ignored.

MarkMichaelis committed:
Added top level statement and refactored PowerShellTestsUtilities

MarkMichaelis committed:
Refactored PowerShellTestsUtilities into a solution shared file.

MarkMichaelis committed:
Minor rename of PowerShellTestUtilities

MarkMichaelis committed:
Excluded "HelloWorld" files 

Also removed top level statement (which wasn't needed in this branch)

BenjaminMichaelis commented on src/Chapter10.Tests/Listing10.13.MakingTypesAvailableOutsideAnAssembly.Tests.cs:
```diff
@@ -14,26 +13,15 @@ public class TypeTests
-            if (PowerShellTestsUtilities.PowerShellNotInstalled()) Assert.Inconclusive("Powershell not installed");
+            if (PowerShellTestUtilities.PowerShellNotInstalled) Assert.Inconclusive("Powershell not installed");
             //EssentialCSharp\\src\\Chapter10.Tests\\bin\\Debug\\netcoreapp3.0
             string ps1Path = Path.GetFullPath("../../../../Chapter10/Listing10.13.MakingTypesAvailableExternally.ps1", Environment.CurrentDirectory);
```
You can use Intellitect.Multitool.RepositoryPaths.GetDefaultRepoRoot() to get the root of the repo

Keboo commented on src/Shared/Tests/PowerShellTestUtilities.cs:
```diff
@@ -0,0 +1,77 @@
+
+        public static Lazy<string?> _PowerShellCommand = new Lazy<string?>(() =>
+        {
+            string? powershellCommand = WindowsEnvironment() ? "powershell" : null;
```
Is there any reason not to use powershell core on all platforms?

Keboo commented on src/Shared/Tests/PowerShellTestUtilities.cs:
```diff
@@ -0,0 +1,77 @@
+
+            if(!System.IO.File.Exists(scriptPath))
+            {
+                throw new InvalidOperationException($"scriptPath, '{scriptPath}', not found.");
```
nit: This should be an `ArgumentException`

Keboo commented on src/Shared/Tests/PowerShellTestUtilities.cs:
```diff
@@ -0,0 +1,77 @@
+            return powerShell.ExitCode;
+        }
+
+        public static string? PowerShellCommand => _PowerShellCommand!.Value;
```
Is the null forgiveness operator needed here? The field is non-nullable.
```suggestion
        public static string? PowerShellCommand => _PowerShellCommand.Value;
```

MarkMichaelis committed:
PowerShell Core by Default

Switched to use PowerShell core by default and to test for PowerShell by executing the PowerShell comment.

BenjaminMichaelis commented on src/Chapter10.Tests/Listing10.23.RegisteringAFinalizerWithProcessExit.Tests.cs:
```diff
@@ -2,62 +2,30 @@
     public class DisposeTests
     {
-
         static string Ps1Path { get; } = Path.GetFullPath("../../../../Chapter10/Listing10.23.RegisteringAFinalizerWithProcessExit.ps1", Environment.CurrentDirectory);
```
```suggestion
        static string Ps1Path { get; } = Path.GetFullPath(Path.Combine("..,..,..,..,Chapter10,Listing10.23.RegisteringAFinalizerWithProcessExit.ps1"), Environment.CurrentDirectory);
```
Shouldn't this use commas to make it cross-plat compatible?

MarkMichaelis committed:
Fixed failing tests (from inattention to detail refactoring)

MarkMichaelis committed:
Minor refactoring

MarkMichaelis committed:
Replaced left over debugging line with non-hard-coded path.

MarkMichaelis committed:
Powershell refactoring

MarkMichaelis committed:
Updated ClassCleanup to use refactored method.

MarkMichaelis commented:
This is no longer active.
